### PR TITLE
python3Packages.hikari: init at 2.0.0.dev124

### PR DESCRIPTION
**Pull request by tomodachi94**
###### Description of changes
Add new package.

[Hikari](https://hikari-py.dev) is an asynchronous Discord API wrapper for Python.

I'm adding this library so I can package other libraries that depend on this package, such as `hikari-lightbulb`/`tanjun`/`crescent`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 232557` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.hikari-python</li>
    <li>python310Packages.hikari-python.dist</li>
    <li>python311Packages.hikari-python</li>
    <li>python311Packages.hikari-python.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/hikari-python/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/hikari-python/default.nix:62]: What is the reason for emptying requirements.txt?

File: pkgs/development/python-modules/hikari-python/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/hikari-python/default.nix:22]: Maybe name it
```suggestion
  pname = "hikari";
```
since it will already be called with `python3Packages.hikari`

**Review by Janik-Haag — CHANGES_REQUESTED** [review]
You should also add 

```nix
pythonImportChecks = [ "hikari" ];
```
and might want to add

```nix
disabled = pythonOlder "3.X";
```

File: pkgs/development/python-modules/hikari-python/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari-python/default.nix:62]: Version constraints were acting oddly with Nixpkgs.

File: pkgs/development/python-modules/hikari-python/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari-python/default.nix:22]: I'm concerned about package name conflicts with the Hikari window manager (`pkgs.hikari`).

Comment (tomodachi94) [reply pkgs/development/python-modules/hikari-python/default.nix]: I've changed the name to plain `hikari` and it doesn't look like anything else broke... my bad!

File: pkgs/development/python-modules/hikari-python/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari-python/default.nix]: @Janik-Haag should I not do this?

File: pkgs/development/python-modules/hikari-python/default.nix
Comment (Janik-Haag) [reply pkgs/development/python-modules/hikari-python/default.nix]: Don't worry, nothing broke because it's when changing it you didn't change it to `pkgs.hikari` you changed it to `python3Packages.hikari` before it was `python3Packages.hikari-python`

File: pkgs/development/python-modules/hikari-python/default.nix
Comment (Janik-Haag) [reply pkgs/development/python-modules/hikari-python/default.nix]: To be honest I'm not quite sure.

Comment (natsukium) [reply pkgs/development/python-modules/hikari-python/default.nix]: We can use `pythonRelaxDepsHooks` here.

File: pkgs/development/python-modules/hikari/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/hikari/default.nix:42]: Would you please add this dependency as `passthru.optional-dependencies`?
see, https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#optional-extra-dependencies-python-optional-dependencies

File: pkgs/development/python-modules/hikari/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/hikari/default.nix:18]:
```suggestion
```
We should not add parameters like this.
It would be better to use `optional-dependencies`.

File: pkgs/development/python-modules/hikari/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/hikari/default.nix:33]:
```suggestion
  nativeBuildInputs = [
```

File: pkgs/development/python-modules/hikari-python/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari-python/default.nix]: Updated to use `pythonRelaxDepsHook`.

File: pkgs/development/python-modules/hikari/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/default.nix]: Moved everything to `nativeCheckInputs`. There is a very specific direct dependency on pytest, so that was included.

File: pkgs/development/python-modules/hikari/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/default.nix]: Fixed.

File: pkgs/development/python-modules/hikari/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/default.nix]: Fixed.

File: pkgs/development/python-modules/hikari/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/default.nix]: Fixed.

File: pkgs/development/python-modules/hikari/default.nix
Comment (Shawn8901) [inline pkgs/development/python-modules/hikari/default.nix:30]: I have some doubts to use leaveDotGit, as it may break on changes upstream as the .git seems to be non deterministic (see https://github.com/NixOS/nixpkgs/issues/8567#issuecomment-1140237191 ) . As far as I can see you are just setting an environment variable with the latest git Sha from the log.
I personally would try to "hardcore" that to the specific Sha of the version (and leave a comment that both need to be updated)
I haven't tested if there are other git commands involved in the build.
Btw: thanks for packaging this, i got an older version in a personal (private) flake and forgot to upstream that. 🙂

Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/default.nix:30]: > [the] best solution would be to have an update script that updates the version and the Sha

Correct me if I'm wrong, but can't update scripts only return a version (no SHAs or other data).

Comment (natsukium) [reply pkgs/development/python-modules/hikari/default.nix:30]: We can freely prepare an `update.sh` in which we get and update "sha1".

Comment (Janik-Haag) [reply pkgs/development/python-modules/hikari/default.nix:30]: [nix-update](https://github.com/Mic92/nix-update) should work fine here.

**Review by Janik-Haag — APPROVED** [review]
LGTM

File: pkgs/development/python-modules/hikari/default.nix
Comment (Shawn8901) [reply pkgs/development/python-modules/hikari/default.nix:30]: I don't see how nix-update can update a custom string aside the rev, hash and the packagers strings. Could you point to the right parameters? 

The issue here is to store the Sha beside the version tag, so replace it in the `about.py` so that there is no dependency to `leaveDotGit` as it may break on pushes.
Another option could be ofc to not use version tag, use the tagged Sha and use this in the about.py too, then a comment to always use Shas would also be enough.

Comment (Janik-Haag) [reply pkgs/development/python-modules/hikari/default.nix:30]: Oh I somehow missed that :see_no_evil:

**Review by Shawn8901 — APPROVED** [review]

**Comment by tomodachi94** [comment]
There are still some minor kinks with this PR, I'm marking this as a draft so it doesn't get merged before I fix them.

**Comment by Janik-Haag** [comment]
@tomodachi94 what are the remaining todos before we are able to get this merged?

**Comment by tomodachi94** [comment]
> @tomodachi94 what are the remaining todos before we are able to get this merged?

I'm not sure at the moment, I have much more time today so I'm taking a look as we speak.

File: pkgs/development/python-modules/hikari/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/default.nix]: I've removed `leaveDotGit` in favor of manually specifying the Git hash.

**Comment by tomodachi94** [comment]
Package now builds and is deterministic. Opening for review.

**Review by Janik-Haag — APPROVED** [review]

File: pkgs/development/python-modules/hikari/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/hikari/default.nix:67]: It seems this environment variable is no longer needed.

File: pkgs/development/python-modules/hikari/README.md
Comment (natsukium) [inline pkgs/development/python-modules/hikari/README.md:5]: Even if `git_hash` is not updated, the build itself will succeed.
This would be a problem with automatic updates using tools like `nixpkgs-update`.
I think getting the version in `update.sh` and set `passthru.update-script` would be better.

File: pkgs/development/python-modules/hikari/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/default.nix:67]: It is needed, see the line immediately after this one. I used a variable for readability.

File: pkgs/development/python-modules/hikari/README.md
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/README.md:5]: Yes, it will succeed, but the build will not be correct. The documentation never said it would fail, only that building the package requires the SHA to be set.

File: pkgs/development/python-modules/hikari/default.nix
Comment (natsukium) [reply pkgs/development/python-modules/hikari/default.nix:67]: Isn't that the nix variable `git_hash`?
I could not find the exported `SHA` either in this derivation or the source code.

File: pkgs/development/python-modules/hikari/README.md
Comment (natsukium) [reply pkgs/development/python-modules/hikari/README.md:5]: I mean, we know that there is a manual operation so that we can update properly.
However, in the typical case where "r-ryantm" updates and someone unaware of this operation approves and merges, the incorrect package is released because the build appears successful even though the "git_hash" is mismatched.
If the update-script is appropriately defined, we can avoid these errors.

Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/README.md:5]: I *think* I did it right. My main concern, however, is that the version number will not be properly bumped.

Comment (natsukium) [reply pkgs/development/python-modules/hikari/README.md:5]: It would be great to update the version in `update.sh`.
I think that this updateScript does not work as expected.

If it is difficult, we can revert this change and keep it semi-automatic.

Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/README.md:5]: > I think that this updateScript does not work as expected.

My other main hesitation was that I don't really know how `updateScript` works and what it does. [The manual section](https://nixos.org/manual/nixpkgs/unstable/#var-passthru-updateScript) feels a little light on the _how_.

Comment (Shawn8901) [reply pkgs/development/python-modules/hikari/README.md:5]: Yeah, bit that's due the fact, as there are several updaters around depending on what to do. I guess here a c&p from another update.sh from nixpkgs that loads a release from GH and your script after that should (not tested) give the expected result. You can run the updater also by your own, thus you can test if a update works properly, drop an older version im the default.nix, run the updater for the package and then check if it's the result you are expecting 
Running the updater should be something like
`nix-shell maintainers/scripts/update.nix --argstr package <exposed package name>` (for the package name I am not 100% sue how that is crafted for python packages in that case). I hope that helps a bit. 🙂

File: pkgs/development/python-modules/hikari/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/default.nix]: Fixed, my bad.

File: pkgs/development/python-modules/hikari/README.md
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/README.md:5]: I modified the existing update script to also update the version. Thank you for your help @natsukiem @Shawn8901!

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 232557` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python310Packages.hikari</li>
    <li>python310Packages.hikari.dist</li>
    <li>python311Packages.hikari</li>
    <li>python311Packages.hikari.dist</li>
  </ul>
</details>

**Review by Janik-Haag — COMMENTED** [review]
Can you update the hash in the fetchfromgithub wrapper? Upstream changed something and now it's diffrent.

File: pkgs/development/python-modules/hikari/update.sh
Comment (Shawn8901) [inline pkgs/development/python-modules/hikari/update.sh:9]:
```suggestion
#!/usr/bin/env nix-shell
#! nix-shell -i bash -p gnused curl jq common-updater-scripts
# Update version
TAGS=$(curl -sL "https://api.github.com/repos/hikari-py/hikari/tags")
VERSION="$(echo $TAGS | jq '.[0].name' --raw-output)"
GIT_HASH=$(echo $TAGS | jq 'map(select(.name == "'$VERSION'")) | .[0].commit.sha' --raw-output)

update-source-version "python3Packages.hikari" "$VERSION" --rev=$GIT_HASH
```

File: pkgs/development/python-modules/hikari/default.nix
Comment (Shawn8901) [inline pkgs/development/python-modules/hikari/default.nix:30]: To remove the double logic, we can also use the git commit here instead of the variable git_hash.
the update-source-version can also set a rev by the update script.
src.ref can then be used by the substitution in patchPhase.

Possibly adding a comment that we are intentionally using a commit hash here could be useful for later.

File: pkgs/development/python-modules/hikari/default.nix
Comment (Shawn8901) [inline pkgs/development/python-modules/hikari/default.nix:67]:
```suggestion
    substituteInPlace hikari/_about.py --replace "__git_sha1__.* = (.*)" ${src.rev}
```

File: pkgs/development/python-modules/hikari/default.nix
Comment (Shawn8901) [inline pkgs/development/python-modules/hikari/default.nix:66]: At least for me the comment is just describing what the postPatch is doing, whilst not giving any additional content. I personally would remove that, but if it helps people i am also okay in keeping that.

File: pkgs/development/python-modules/hikari/README.md
Comment (Shawn8901) [inline pkgs/development/python-modules/hikari/README.md:3]: If the suggestion from the update.sh are implemented at least on my opinion we could remove the readme, as we are not giving any needed information here.

Updates for the package should be done by `nix-shell maintainers/scripts/update.nix --argstr package python3Packages.hikari` and with my suggestions that update.sh does all the handling

**Review by Shawn8901 — CHANGES_REQUESTED** [review]

**Comment by natsukium** [comment]
If the points currently noted are fixed, it looks fine to me.

File: pkgs/development/python-modules/hikari/update.sh
Comment (Shawn8901) [reply pkgs/development/python-modules/hikari/update.sh:9]: I found another neat solution to get the commit hash without breaking reproducibility on the nix layer. The pattern used on fastly is to fetch the full git, store the hash and then remove the git folder.

https://github.com/NixOS/nixpkgs/blob/dec7d29544716eadebcc0959c846651fd3b09781/pkgs/misc/fastly/default.nix#L25

Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/update.sh]: I have borrowed the code from the Fastly package and adjusted it for this one; everything works as expected, without an update script being involved.

File: pkgs/development/python-modules/hikari/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/default.nix]: Comment removed.

File: pkgs/development/python-modules/hikari/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/default.nix]: Unused in latest push.

File: pkgs/development/python-modules/hikari/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/default.nix:25]: See https://github.com/NixOS/nixpkgs/pull/232557#discussion_r1542525480

File: pkgs/development/python-modules/hikari/default.nix
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/default.nix]: Removed, thanks for catching this.

File: pkgs/development/python-modules/hikari/README.md
Comment (tomodachi94) [reply pkgs/development/python-modules/hikari/README.md]: Unused, see https://github.com/NixOS/nixpkgs/pull/232557#discussion_r1542525480

**Comment by tomodachi94** [comment]
This new revision of the PR removes the complex update logic in favor of the `fastly` method [mentioned here](https://github.com/NixOS/nixpkgs/pull/232557#discussion_r1271327627). As a result, this package is slightly simpler to maintain.

**Review by Shawn8901 — APPROVED** [review]
lgtm

**Review by jshcmpbll — APPROVED** [review]
LGTM

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**
